xdsl/dialects/tosa.py

```diff
@@ -68,11 +70,71 @@ class RescaleOp(IRDLOperation):
     assembly_format = "$input attr-dict `:` `(` type($input) `)` `->` type($output)"
 
 
+@irdl_op_definition
+class AddOp(IRDLOperation):
+    """
+    Tosa elementwise add operation
+    """
+
+    name = "tosa.add"
+
+    in1 = operand_def(TensorType)
+    in2 = operand_def(TensorType)
+    output = result_def(TensorType)
+
+    assembly_format = "$in1 `,` $in2 attr-dict `:` `(` type($in1) `,` type($in2) `)` `->` type($output)"
+
+    def verify_(self) -> None:
+        """
+        Verify that the two input tensors are compatible, and that the result type can be constructed. For this,
+        both tensors must have the same rank. They should either have the same number of elements per dim, or
+        if there is only one element it can be broadcast implcitly.
+        """
+        t1 = self.in1.type
+        t2 = self.in2.type
+        t_out = self.output.type
+
+        assert isa(t1, TensorType)
+        assert isa(t2, TensorType)
+        assert isa(t_out, TensorType)
+        
+        if not (t1.get_num_dims() == t2.get_num_dims() == t_out.get_num_dims()):
+            raise VerifyException(
+                "'tosa.add' rank mismatch between input and output tensors"
+            )
+
+        if not (
+            t1.get_element_type() == t2.get_element_type() == t_out.get_element_type()
+        ):
+            raise VerifyException(
+                "'tosa.add' element type mismatch between inputs and output tensors"
+            )
```

Comment:
Can you please move this into the constraint of the operation? Ideally with a `VarConstraint` and `TensorType.constr`